net/socks5/socks5.go

```diff
@@ -295,123 +488,202 @@ func parseClientAuth(r io.Reader) (usr, pwd string, err error) {
+	case ipv6:
+		addr = net.ParseIP(s.addr).To16()
+		if addr == nil {
+			return nil, fmt.Errorf("invalid IPv6 address for binding")
+		}
+	default:
+		return nil, fmt.Errorf("unsupported address type")
+	}
+
+	pkt := []byte{byte(s.addrType)}
+	pkt = append(pkt, addr...)
+	pkt = binary.BigEndian.AppendUint16(pkt, s.port)
+	return pkt, nil
+}
+func (s socksAddr) hostPort() string {
+	return net.JoinHostPort(s.addr, strconv.Itoa(int(s.port)))
+}
 
 // response contains the contents of
 // a response packet sent from the proxy
 // to the client.
 type response struct {
-	reply        replyCode
-	bindAddrType addrType
-	bindAddr     string
-	bindPort     uint16
+	reply    replyCode
+	bindAddr socksAddr
+}
+
+func errorResponse(code replyCode) *response {
+	return &response{reply: code, bindAddr: zeroSocksAddr}
 }
 
 // marshal converts a SOCKS5Response struct into
 // a packet. If res.reply == Success, it may throw an error on
 // receiving an invalid bind address. Otherwise, it will not throw.
 func (res *response) marshal() ([]byte, error) {
-	pkt := make([]byte, 4)
+	pkt := make([]byte, 3)
 	pkt[0] = socks5Version
 	pkt[1] = byte(res.reply)
 	pkt[2] = 0 // null reserved byte
-	pkt[3] = byte(res.bindAddrType)
 
-	if res.reply != success {
-		return pkt, nil
+	addrPkt, err := res.bindAddr.marshal()
+	if err != nil {
+		return nil, err
 	}
 
-	var addr []byte
-	switch res.bindAddrType {
-	case ipv4:
-		addr = net.ParseIP(res.bindAddr).To4()
-		if addr == nil {
-			return nil, fmt.Errorf("invalid IPv4 address for binding")
-		}
-	case domainName:
-		if len(res.bindAddr) > 255 {
-			return nil, fmt.Errorf("invalid domain name for binding")
-		}
-		addr = make([]byte, 0, len(res.bindAddr)+1)
-		addr = append(addr, byte(len(res.bindAddr)))
-		addr = append(addr, []byte(res.bindAddr)...)
-	case ipv6:
-		addr = net.ParseIP(res.bindAddr).To16()
-		if addr == nil {
-			return nil, fmt.Errorf("invalid IPv6 address for binding")
-		}
-	default:
-		return nil, fmt.Errorf("unsupported address type")
+	return append(pkt, addrPkt...), nil
+}
+
+type udpRequest struct {
+	frag byte
+	addr socksAddr
+}
+
+// +----+------+------+----------+----------+----------+
+// |RSV | FRAG | ATYP | DST.ADDR | DST.PORT |   DATA   |
+// +----+------+------+----------+----------+----------+
+// | 2  |  1   |  1   | Variable |    2     | Variable |
+// +----+------+------+----------+----------+----------+
+func parseUdpRequest(data []byte) (*udpRequest, []byte, error) {
+	if len(data) < 4 {
+		return nil, nil, fmt.Errorf("invalid packet length")
 	}
 
-	pkt = append(pkt, addr...)
-	pkt = binary.BigEndian.AppendUint16(pkt, uint16(res.bindPort))
+	// reserved bytes
+	if !(data[0] == 0 && data[1] == 0) {
+		return nil, nil, fmt.Errorf("invalid udp request header")
+	}
 
-	return pkt, nil
+	frag := data[2]
+
+	reader := bytes.NewReader(data[3:])
+	addr, err := parseSocksAddr(reader)
+	body, _ := io.ReadAll(reader)
+	return &udpRequest{
+		frag: frag,
+		addr: addr,
+	}, body, err
```

Comment:
this looks like a lot of allocations to do per packet. UDP is heavily used by QUIC nowadays and isn't just occasional DNS queries anymore. We should probably care about allocs/throughput a bit.

If this func gets the whole payload in `data []byte`, I don't know why we need to go to an io.Reader and back out (via io.ReadAll).

We should just have parseSocksAddr work in terms of []byte?

Reply:
The reason parseSocksAddr uses an io.Reader is due to its need to handle variable-length address fields dynamically. Since the length and format of the address can vary (e.g., IPv4, IPv6, or a domain name with variable length), using an io.Reader abstracts the complexity of handling these different formats efficiently.

Is it okay if I want to fix it like this?
```golang
bodyLen := reader.Len() // (*bytes.Reader).Len() return unread data length
body := data[len(data)-bodyLen:]
```

Reply:
By the way, this is my first time submitting a PR to Tailscale. Should I fix the code review issues in the previous commit and force push, or should I make a new commit for these changes?